File: src/__tests__/frontmatter.test.ts
import { describe, it, expect } from 'vitest'
import {
  parseFrontmatter,
  stampFrontmatter,
  frontmatterRange,
  permalinkForLogPath,
  isDraftPath,
  isPublishedPath,
} from '@/utils/frontmatter'

const SAMPLE = `---
title: My Post
public: false
date: 2026-06-08 10:00:00
post: true
tags:
  - writing
  - dev
---

Hello world.
`

describe('parseFrontmatter', () => {
  it('parses standard fields', () => {
    const { fm, body } = parseFrontmatter(SAMPLE)
    expect(fm.title).toBe('My Post')
    expect(fm.public).toBe(false)
    expect(fm.post).toBe(true)
    expect(fm.date).toBe('2026-06-08 10:00:00')
    expect(body).toContain('Hello world.')
  })

  it('parses block-list tags', () => {
    const { fm } = parseFrontmatter(SAMPLE)
    expect(fm.tags).toEqual(['writing', 'dev'])
  })

  it('parses inline-array tags', () => {
    const { fm } = parseFrontmatter('---\ntags: [a, b]\n---\nx')
    expect(fm.tags).toEqual(['a', 'b'])
  })

  it('parses scalar tags (single + comma list)', () => {
    expect(parseFrontmatter('---\ntags: solo\n---\n').fm.tags).toEqual(['solo'])
    expect(parseFrontmatter('---\ntags: a, b\n---\n').fm.tags).toEqual(['a', 'b'])
  })

  it('returns empty fm for content without frontmatter', () => {
    const { fm, body } = parseFrontmatter('just text')
    expect(fm).toEqual({})
    expect(body).toBe('just text')
  })

  it('strips quotes from values', () => {
    const { fm } = parseFrontmatter('---\ntitle: "Quoted"\n---\n')
    expect(fm.title).toBe('Quoted')
  })
})

describe('stampFrontmatter', () => {
  it('replaces an existing scalar key', () => {
    const out = stampFrontmatter(SAMPLE, { title: 'New Title' })
    expect(parseFrontmatter(out).fm.title).toBe('New Title')
    // Body untouched
    expect(out).toContain('Hello world.')
  })

  it('appends a missing key', () => {
    const out = stampFrontmatter(SAMPLE, { project: 'memo' })
    expect(parseFrontmatter(out).fm.project).toBe('memo')
  })

  it('replaces block-list tags with new block list', () => {
    const out = stampFrontmatter(SAMPLE, { tags: ['x', 'y', 'z'] })
    expect(parseFrontmatter(out).fm.tags).toEqual(['x', 'y', 'z'])
    // No leftover old items
    expect(out).not.toContain('- writing')
  })

  it('replaces scalar tags with block list', () => {
    const src = '---\ntags: old\n---\nbody'
    const out = stampFrontmatter(src, { tags: ['new1', 'new2'] })
    expect(parseFrontmatter(out).fm.tags).toEqual(['new1', 'new2'])
    expect(out).not.toContain('tags: old')
  })

  it('empty tags array leaves an empty tags key', () => {
    const out = stampFrontmatter(SAMPLE, { tags: [] })
    expect(parseFrontmatter(out).fm.tags).toBeUndefined()
    expect(out).toContain('tags: ')
  })

  it('round-trips: parse(stamp(x)) preserves untouched keys', () => {
    const out = stampFrontmatter(SAMPLE, { title: 'Changed' })
    const { fm } = parseFrontmatter(out)
    expect(fm.public).toBe(false)
    expect(fm.post).toBe(true)
    expect(fm.tags).toEqual(['writing', 'dev'])
    expect(fm.date).toBe('2026-06-08 10:00:00')
  })

  it('creates frontmatter on content without any', () => {
    const out = stampFrontmatter('plain body', { title: 'T' })
    const { fm, body } = parseFrontmatter(out)
    expect(fm.title).toBe('T')
    expect(body).toBe('plain body')
  })
})

describe('frontmatterRange', () => {
  it('covers the full fence block', () => {
    const r = frontmatterRange(SAMPLE)!
    expect(r.from).toBe(0)
    expect(SAMPLE.slice(r.from, r.to)).toMatch(/^---\n[\s\S]*\n---\n$/)
  })

  it('null when no frontmatter', () => {
    expect(frontmatterRange('nope')).toBeNull()
  })
})

describe('permalinkForLogPath', () => {
  it('maps log path to public URL', () => {
    expect(permalinkForLogPath('log/2026-06-08-10-00-00.md'))
      .toBe('https://yousefamar.com/memo/log/2026-06-08-10-00-00/')
  })

  it('null for non-log paths', () => {
    expect(permalinkForLogPath('scratch/blog-drafts/foo.md')).toBeNull()
    expect(permalinkForLogPath('projects/al/index.md')).toBeNull()
  })
})

describe('path classifiers', () => {
  it('isDraftPath', () => {
    expect(isDraftPath('scratch/blog-drafts/x.md')).toBe(true)
    expect(isDraftPath('log/x.md')).toBe(false)
    expect(isDraftPath(null)).toBe(false)
  })

  it('isPublishedPath', () => {
    expect(isPublishedPath('log/x.md')).toBe(true)
    expect(isPublishedPath('log/sub/x.md')).toBe(false)
    expect(isPublishedPath('scratch/blog-drafts/x.md')).toBe(false)
  })
})

File: android/app/src/test/kotlin/io/amar/console/data/notes/FrontmatterTest.kt
package io.amar.console.data.notes

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class FrontmatterTest {

    @Test
    fun `parses title date project and scalar tags`() {
        val fm = FrontmatterParser.parse(
            """
            ---
            title: My Post
            date: 2026-07-20
            project: console
            tags: foo, bar
            ---
            body here
            """.trimIndent()
        )
        assertEquals("My Post", fm.title)
        assertEquals("2026-07-20", fm.date)
        assertEquals("console", fm.project)
        assertEquals(listOf("foo", "bar"), fm.tags)
    }

    @Test
    fun `parses inline array and block-list tags`() {
        assertEquals(
            listOf("a", "b"),
            FrontmatterParser.parse("---\ntags: [a, b]\n---\n").tags,
        )
        assertEquals(
            listOf("x", "y"),
            FrontmatterParser.parse("---\ntags:\n  - x\n  - y\n---\n").tags,
        )
    }

    @Test
    fun `no frontmatter yields empty`() {
        val fm = FrontmatterParser.parse("just a body\nno fences")
        assertNull(fm.title)
        assertTrue(fm.tags.isEmpty())
    }

    @Test
    fun `stamp replaces title in place preserving body`() {
        val out = FrontmatterParser.stamp("---\ntitle: Old\ndate: x\n---\nbody", listOf("title" to "New"))
        val fm = FrontmatterParser.parse(out)
        assertEquals("New", fm.title)
        assertEquals("x", fm.date)
        assertTrue(out.endsWith("body"))
    }

    @Test
    fun `stamp appends missing key`() {
        val out = FrontmatterParser.stamp("---\ntitle: T\n---\nbody", listOf("project" to "p"))
        assertEquals("p", FrontmatterParser.parse(out).project)
    }

    @Test
    fun `stamp tags as block list replacing scalar form`() {
        val out = FrontmatterParser.stamp("---\ntags: old\ntitle: T\n---\nbody", listOf("tags" to listOf("a", "b")))
        assertEquals(listOf("a", "b"), FrontmatterParser.parse(out).tags)
        assertEquals("T", FrontmatterParser.parse(out).title)
    }

    @Test
    fun `stamp creates frontmatter when absent`() {
        val out = FrontmatterParser.stamp("plain body", listOf("title" to "T"))
        assertTrue(out.startsWith("---\n"))
        assertEquals("T", FrontmatterParser.parse(out).title)
        assertTrue(out.endsWith("plain body"))
    }

    @Test
    fun `range covers the fenced block`() {
        val content = "---\ntitle: T\n---\nbody"
        val r = FrontmatterParser.range(content)!!
        assertEquals(0, r.first)
        assertEquals("---\ntitle: T\n---\n", content.substring(r.first, r.last + 1))
        assertNull(FrontmatterParser.range("no fm"))
    }

    @Test
    fun `permalink and path detection`() {
        assertEquals(
            "https://yousefamar.com/memo/log/2026-07-20-hello/",
            FrontmatterParser.permalinkForLogPath("log/2026-07-20-hello.md"),
        )
        assertEquals(
            "https://yousefamar.com/memo/log/2026-07-20-hello/",
            FrontmatterParser.permalinkForLogPath("projects/console/log/2026-07-20-hello.md"),
        )
        assertNull(FrontmatterParser.permalinkForLogPath("scratch/foo.md"))
        assertTrue(FrontmatterParser.isDraftPath("scratch/blog-drafts/idea.md"))
        assertTrue(FrontmatterParser.isPublishedPath("log/x.md"))
        assertTrue(FrontmatterParser.isPublishedPath("projects/console/log/x.md"))
        assertFalse(FrontmatterParser.isPublishedPath("projects/console/log/drafts/x.md"))
        assertFalse(FrontmatterParser.isPublishedPath("log/sub/x.md"))
        assertTrue(FrontmatterParser.isWritingFile("log/x.md"))
    }

    @Test
    fun `yamlScalar quotes what a plain scalar cannot hold`() {
        assertEquals("\"Foo: bar\"", FrontmatterParser.yamlScalar("Foo: bar"))
        assertEquals("\"Wait for it:\"", FrontmatterParser.yamlScalar("Wait for it:"))
        assertEquals("\"- dashed\"", FrontmatterParser.yamlScalar("- dashed"))
        assertEquals("\"#hash\"", FrontmatterParser.yamlScalar("#hash"))
        assertEquals("\"a #b\"", FrontmatterParser.yamlScalar("a #b"))
        assertEquals("\" padded \"", FrontmatterParser.yamlScalar(" padded "))
        assertEquals("\"\"", FrontmatterParser.yamlScalar(""))
        // Would come back as a non-string
        assertEquals("\"true\"", FrontmatterParser.yamlScalar("true"))
        assertEquals("\"2026\"", FrontmatterParser.yamlScalar("2026"))
        assertEquals("\"0x10\"", FrontmatterParser.yamlScalar("0x10"))
        assertEquals("\"~\"", FrontmatterParser.yamlScalar("~"))
        // Escapes
        assertEquals("\"He said \\\"hi\\\": ok\"", FrontmatterParser.yamlScalar("He said \"hi\": ok"))
        assertEquals("\"two: lines\\nhere\"", FrontmatterParser.yamlScalar("two: lines\nhere"))
        // Already safe — no churn
        assertEquals("plain title", FrontmatterParser.yamlScalar("plain title"))
        assertEquals("Sainsbury's CLI", FrontmatterParser.yamlScalar("Sainsbury's CLI"))
        assertEquals("2026-06-08 10:00:00", FrontmatterParser.yamlScalar("2026-06-08 10:00:00"))
        // Non-strings pass through
        assertEquals("true", FrontmatterParser.yamlScalar(true))
        assertEquals("7", FrontmatterParser.yamlScalar(7))
    }

    @Test
    fun `yamlScalar round-trips through unquote`() {
        val hostile = listOf(
            "Foo: bar", "Wait for it:", "plain", "Sainsbury's CLI",
            "He said \"hi\": really", "Bruteforcing tailnet \"fun names\"",
            "true", "2026", "back\\slash: x", "tab:\there", " padded ", "",
        )
        for (s in hostile) assertEquals(s, FrontmatterParser.unquote(FrontmatterParser.yamlScalar(s)))
    }

    @Test
    fun `unquote only strips a matched pair`() {
        assertEquals(
            "Bruteforcing tailnet \"fun names\"",
            FrontmatterParser.parse("---\ntitle: Bruteforcing tailnet \"fun names\"\n---\n").title,
        )
        assertEquals(
            "Blast: \"Cult II\" it's",
            FrontmatterParser.parse("---\ntitle: 'Blast: \"Cult II\" it''s'\n---\n").title,
        )
    }

    @Test
    fun `a quoted tags scalar is one tag`() {
        assertEquals(listOf("a, b"), FrontmatterParser.parse("---\ntags: \"a, b\"\n---\n").tags)
    }

    @Test
    fun `stamp escapes a title with a colon and tags with hazards`() {
        val out = FrontmatterParser.stamp(
            "---\ntitle: Old\n---\nbody",
            listOf("title" to "Foo: bar", "tags" to listOf("a: b", "plain")),
        )
        assertTrue(out.contains("title: \"Foo: bar\""))
        assertTrue(out.contains("  - \"a: b\""))
        val fm = FrontmatterParser.parse(out)
        assertEquals("Foo: bar", fm.title)
        assertEquals(listOf("a: b", "plain"), fm.tags)
    }

    @Test
    fun `project slug helpers`() {
        assertEquals("console", FrontmatterParser.projectSlugFromPath("projects/console.md"))
        assertEquals("console", FrontmatterParser.projectSlugFromPath("projects/console/index.md"))
        assertEquals("console", FrontmatterParser.enclosingProjectSlug("projects/console/devlog/1.md"))
        assertNull(FrontmatterParser.enclosingProjectSlug("scratch/x.md"))
    }
}
